docs: 0.2.0 Changelog (#334)
Unity-Technologies/com.unity.multiplayer.samples.coop: Packages/com.unity.multiplayer.samples.coop/CHANGELOG.md

  * The UI now tracks and displays player has arrived, number of connected players in the lobby, and player status. 

* Added F/X and animation assets for the game including:

  * Audio files for boss sound effects
  * Ramp-up animation for hero movement and actions
  * Visual effects and particles for Tank charge skill 
  * Art assets to wave spawner, including animations for ReceiveDamage, Broken (died), and Revive 
  * Game music 

* Added audiomixer to Boss Room for separate audio channels and a master mixer. 

* Updated and added various hero abilities:

  * Added a cooldown to Archer's PowerShot. 
  * Added the Rogue's Dagger and Dash skills. 
  * Added the Rogue's Sneak skill using local stealth, applying a graphical effect to the stealthy character while still sending all network events normally.
  * Properly display Heal abilities when targeting a fallen ally character. 
  * Character attack actions properly support Hold to charge options. 

* To show how UI elements and game objects can be networked, added networked functionality using `INetworkSerializable` in the `CharSelect` screen to network player's selected character on the Character Selection screen. 

> * Added a Photon filter to remove any invisible characters introduced by Zoom when copy-pasting room names. 

Reviewer: ```suggestion
* Added a filter to remove any invisible characters introduced by Zoom when copy-pasting room names, IP addresses, or ports. 
```

* Removed a previous work-around for character selections when host replays a completed game. The issue was resolved, allowing players to see character selections during replay. 
* Fixed collision wall settings, fixing an issues where the boss knock-back ability sent players through walls.
* Resolved an issue where any players leaving the lobby sent all players to the lobby.
* Fixed animations for enemies including the smoke animation for destroyed imps and the boss helmet when crying.

### Known issues

* An MLAPI soft sync error on cleanup between scene transitions may break the game, for example imps do not spawn and pots are intangible.
* The game can be initiated while a second player is connecting to the host in `CharSelect`. Players may join without selected characters spawning and in an unresponsive state.
* Sometimes after completing a match and the host starts a new match from the Victory or Loss screen, connected players may have no visible interactions to join or select characters. A work-around is implemented to not block entry into the game. 

## [0.1.0] - 2021-04-07

v0.1.0 is an Early Access release for Multiplayer Samples Co-op.

> It requires and supports Unity 2020.3.0f1 and Unity MLAPI v0.1.0. For additional information on MLAPI, see the changelog and release notes for the Unity MLAPI package.

Reviewer 2: ```suggestion
It requires and supports Unity 2020.3.8f1 and Unity MLAPI v0.1.0. For additional information on MLAPI, see the changelog and release notes for the Unity MLAPI package.
```

    * Added two separator `GameObject`s for scene readability: runtime `NetworkObject`s and `NetworkObject`s already placed in the scene.
    * Added a custom editor for GameEvents to fire in the editor (greatly enhances testing).
    * The `LifeState` NetworkVariable was moved from `NetworkCharacterState` into its own component, `NetworkLifeState`.
    * Cleaned up and removed old spawn prefab collections and spawner scripts (`NetSpawnPoint`).

* Added ramp-up animation for hero movement and actions
* Added F/X and animation assets for the game including:

  * Audio files for boss sound effects
  * Visual effects and particles for Tank charge skill 
  * Art assets to wave spawner, including animations for ReceiveDamage, Broken (died), and Revive 

>   * Game music 

Reviewer: ```suggestion
```

> * Added audiomixer to Boss Room for separate audio channels and a master mixer. 

Reviewer: ```suggestion
```

* Reviewed and revised code to better following quality standards.
* Updated the Mage character base attack to better support the new enqueuing ability and handle game behaviors. Updates include:

  * Actions being non-blocking now allow other actions while a mage-bolt is in flight
  * Actions send in `ActionSequence` groups to better handle actions when players spam click enemies
  * Timing issues with animations, actions, and character location
  * Bolt animation for legitimate hits
  * Updated attacks not to knock rogues out of stealth 

* Updated character attacks to not cause friendly-fire damage between players. 
* Updated and resolved issues with 3D models including polygon count of coins and chest, artifacts where level graphics are stitched together or overlapping characters, and asset map consistency used by objects (color + normal map + emission).
* Merged `ConnectStatus` and `DisconnectReason` into a single `ConnectStatus`. 
* Updated `ServerGameNetPortal` to properly handle the following:

  * Per-connection state on client disconnect
  * Additional errors including a full server and player ID (GUID) already playing

> * Updated code to allow hosts to specify a port to listen to, removing the hard-coded port. 

Reviewer: This was done in 0.1.

    * Added two separator `GameObject`s for scene readability: runtime `NetworkObject`s and `NetworkObject`s already placed in the scene.
    * Added a custom editor for GameEvents to fire in the editor (greatly enhances testing).
    * The `LifeState` NetworkVariable was moved from `NetworkCharacterState` into its own component, `NetworkLifeState`.
    * Cleaned up and removed old spawn prefab collections and spawner scripts (`NetSpawnPoint`).

* Updated the user interface including the following:

  * When joining a game, a "Connecting..." UI loads. When disconnecting from a game, you are returned to the MainMenuScene with a "Connection to Host lost" message. If the game fails to connect, a general message "Connection to Host failed" loads. 
  * Added an option to leave the lobby after joining it. You no longer have to restart the game to leave the lobby.

>   * Added option to cancel the game after clicking **Ready**. 

Reviewer 2: ```suggestion
  * Added option to cancel my selection after clicking **Ready**. 
```

  * Added a gear icon for accessing and modifying app settings. 
  * The UI now tracks and displays player has arrived, number of connected players in the lobby, and player status. 

* Added F/X and animation assets for the game including:

  * Audio files for boss sound effects
  * Ramp-up animation for hero movement and actions
  * Visual effects and particles for Tank charge skill 
  * Art assets to wave spawner, including animations for ReceiveDamage, Broken (died), and Revive 
  * Game music 

* Added audiomixer to Boss Room for separate audio channels and a master mixer. 

* Updated and added various hero abilities:

  * Added a cooldown to Archer's PowerShot. 

>   * Added the Rogue's Dagger and Dash skills. 

Reviewer 2: ```suggestion
  * Added the Rogue's Dagger and Dash skills. The dash skill shows an install teleport (using an RPC) instead of a charge like the boss' (which updates its position over time). 
```

  * Added the Rogue's Sneak skill using local stealth, applying a graphical effect to the stealthy character while still sending all network events normally.
  * Properly display Heal abilities when targeting a fallen ally character. 
  * Character attack actions properly support Hold to charge options. 

* To show how UI elements and game objects can be networked, added networked functionality using `INetworkSerializable` in the `CharSelect` screen to network player's selected character on the Character Selection screen. 
* Added a filter to remove any invisible characters introduced by Zoom when copy-pasting room names, IP addresses, or ports. 

> * Boss Room now uses the [UnityToonShader](https://github.com/IronWarrior/UnityToonShader) for rendering the 3D surfaces to emulate 2D, flat surfaces.

Reviewer 2: That's not new to v0.2, that's been there for a while

Author: Gotcha was in the tickets, will pull

> * Boss Room now uses the [UnityToonShader](https://github.com/IronWarrior/UnityToonShader) for rendering the 3D surfaces to emulate 2D, flat surfaces.

Reviewer 2: ```suggestion

```

> * Boss Room now uses the [UnityToonShader](https://github.com/IronWarrior/UnityToonShader) for rendering the 3D surfaces to emulate 2D, flat surfaces.

Author: ```suggestion

```

> * Added disconnection error message to load when a player or host disconnects due to limited or no network connectivity. Client logic was also updated to detect Host disconnection scenarios, such as losting connectivity.

Reviewer 2: ```suggestion
* Added disconnection error message to load when a player or host disconnects. Client logic was also updated to detect Host disconnection scenarios, such as losting connectivity.
```

Reviewer 2: we don't support bandwidth issue detection right now

Author: Sounds good! Thanks!

* Balanced hero and enemy stats, spawners, trigger areas, and enemy detetction areas. 
* Added healthbars to display when damaged and properly track imp health locally and across clients. 

### Changes

* Updated Boss Room per asset and project settings for increased performance and better resource management:

  * Disabled GPU Skinning to optimize GPU usage and FPS.
  * Lowered quality of ambient occlusion from high to medium.
  * Switched SMAA High to FXAA (fast mode) reducing GPU cost.
  * Modified GPU Instancing on imps, heroes, and the boss to significantly reduce the number of draw calls.
  * Turned off Cast Shadows on Imp and Imp Boss.
  * Disabled mesh colliders of lava, which is more decorative than interactive.

>   * Refactored the S_SimpleDissolve shader which consumed most import time.

Reviewer 2: This was already in v0.1
I'm not sure about the other ones, @fernando-cortez can you make sure?

Reviewer: The ones listed on here are correct, since they went in recently for the PR review change. I'll take a look at the rest of the changes.

> * Updated the Photon Setup Guide, indicating you need only app ID when playing with friends. For users connecting across regions, you may need to hard code a region in your app settings by using the room code and region instead of just the room code sharing in game. 

Reviewer 2: mmm where is that coming from? Right now, regions are hardcoded in the room name you're sharing

Author: Hmm this was a ticket GOMPS-88, can remove this one if we want!

* Reviewed and revised code to better following quality standards.
* Updated the Mage character base attack to better support the new enqueuing ability and handle game behaviors. Updates include:

  * Actions being non-blocking now allow other actions while a mage-bolt is in flight
  * Actions send in `ActionSequence` groups to better handle actions when players spam click enemies
  * Timing issues with animations, actions, and character location
  * Bolt animation for legitimate hits
  * Updated attacks not to knock rogues out of stealth 

* Updated character attacks to not cause friendly-fire damage between players. 
* Updated and resolved issues with 3D models including polygon count of coins and chest, artifacts where level graphics are stitched together or overlapping characters, and asset map consistency used by objects (color + normal map + emission).
* Merged `ConnectStatus` and `DisconnectReason` into a single `ConnectStatus`. 
* Updated `ServerGameNetPortal` to properly handle the following:

  * Per-connection state on client disconnect
  * Additional errors including a full server and player ID (GUID) already playing

>   * Player connection if they experience a game-level connection failure

Reviewer 2: This is a duplicate from a previous entry saying we handle connection failures

Author: Pulling thanks!
```suggestion

```

* Updated code to allow hosts to specify a port to listen to, removing the hard-coded port. 
* Refactored Action Bar code including the following:

  * Removed the `ButtonID` from `UIHudButton`.
  * Removed hard-coded values from `HeroActionBar`.
  * Removed switch statements.
  * Completed minor code cleanup.
  * Added verification to only show skill and ability buttons for available character abilities. Empty buttons no longer load for characters.

* Added a call to warm up shaders when the project starts to ensure animations issues do not occur.
* Removed collision from objects that have a Broken (dead) state.
* Implemented a better cooldown solution and calculations for tracking and managing character, imp, and boss actions. 
* Updated event registration and unregistration code to be symmetrical across the project.

> ### Fixes

Reviewer 2: It feels like a lot of the above are fixes too.

Author: Will look them over for best location as update vs fix, thanks!!

# Multiplayer Samples Co-op Changelog

## [0.2.0] - 2021-05-18

v0.2.0 is an Early Access release for Multiplayer Samples Co-op.

> It requires and supports Unity v2020.3 and later and Unity MLAPI v0.1.0. For additional information on MLAPI, see the changelog and release notes for the Unity MLAPI package.

Reviewer 3: Thinking we should list the actual version tested with 2020.3.8 LTS - also reword the "and later" because we haven't tested against that, but it could work

Author: Updating versions.

* Added F/X and animation assets for the game including:

  * Audio files for boss sound effects
  * Ramp-up animation for hero movement and actions
  * Visual effects and particles for Tank charge skill 
  * Art assets to wave spawner, including animations for ReceiveDamage, Broken (died), and Revive 
  * Game music 

* Added audiomixer to Boss Room for separate audio channels and a master mixer. 

* Updated and added various hero abilities:

  * Added a cooldown to Archer's PowerShot. 
  * Used generic actions to implement Rogue's Dagger skill. 
  * Used generic actions to implement Rogue's Sneak skill using local stealth, applying a graphical effect to the stealthy character while still sending all network events normally. 
  * Used generic actions to implement Rogue's Dash skill. 
  * Properly display Heal abilities when targeting a fallen ally character. 
  * Character attack actions properly support Hold to charge options. 

* To show how UI elements and game objects can be networked, added networked functionality using `INetworkSerializable` for UI elements on the Character Selection screen including networked mouse cursors. 

> * Added a Photon filter for host room names not to allow profanity, racial slurs, and additionsl socially acceptable words. 

Author: Fixing!

  * Actions being non-blocking allowing other actions while a mage-bolt is in flight
  * Sending actions in `ActionSequence` groups to better handle actions when players spam click enemies
  * Timing issues with animations, actions, and character location
  * Bolt animation for legitimate hits
  * Updated attacks not to knock rogues out of stealth 

* Updated character attacks to not cause friendly-fire damage between players. 
* Updated and resolved issues with 3D models including polygon count of coins and chest, artifacts where level graphics are stitched together or overlapping characters, and asset map consistency used by objects (color + normal map + emission).
* Merged `ConnectStatus` and `DisconnectReason` into a single `ConnectStatus`. 
* Updated `ServerGameNetPortal` to properly handle the following:

  * Per-connection state on client disconnect
  * Additional errors including a full server and player ID (GUID) already playing
  * Player connection if they experience a game-level connection failure

* Updated code to allow hosts to specify a port to listen to, removing the hard-coded port. 

> * Updated the Boss Room UI to only show skill and ability buttons for character abilities. Empty buttons no longer load for characters. 

Author: Thanks! Getting that added!

* Green quads no longer show on impact when the Archer arrow strikes enemies. 
* Fixed issue to correctly allow one player to receive a character when two players in the Character Select click **Ready** for the same hero at the same time. Character Select is no longer blocked.  
* Fixed an issue with boss collisions with a Pillar correctly applying a stun effect and shattering the pillar when using the Trample attack. 
* Fixed the lobby welcome messages to correctly list the player names, including a previous issues for P1 and P2. 
* On Windows, investigated and fixed issues with visible effects for character actions including Mage Freeze attack and Rogue Dash.
* On Wizards, fixed issue with imp spawners not respawning new imps after exploring the room.
* Fixed an issue where the door state does not reflect the existing state when players connect late to a game, for example if other players open the door and a player joins late the door displays as closed. 
* Removed a previous work-around for character selections when host replays a completed game. The issue was resolved, allowing players to see character selections during replay. 

> * Fixed a Null reference for `RoomNameBox` that potentially caused a "NullReferenceExceltion: Object reference not set to an instance of an object".

Author: I added what seemed important. If it shouldn't, I don't mind removing. :)

# Multiplayer Samples Co-op Changelog

## [0.2.0] - 2021-05-18

v0.2.0 is an Early Access release for Multiplayer Samples Co-op.

It requires and supports Unity v2020.3 and later and Unity MLAPI v0.1.0. For additional information on MLAPI, see the changelog and release notes for the Unity MLAPI package.

### New features

> 

Reviewer: Shouldn't we have one here letting users know that we've introduced static scene NetworkObjects? This may be the biggest improvement on the MLAPI side of things.

Author: Will double check tickets for one on this! Thanks!

Reviewer: Yup. Feel free to align on this one offline if needed.

  * Added option to cancel the game after clicking **Ready**. 
  * Added a gear icon for accessing and modifying app settings. 
  * The UI now tracks and displays player has arrived, number of connected players in the lobby, and player status. 

* Added F/X and animation assets for the game including:

  * Audio files for boss sound effects
  * Ramp-up animation for hero movement and actions
  * Visual effects and particles for Tank charge skill 
  * Art assets to wave spawner, including animations for ReceiveDamage, Broken (died), and Revive 
  * Game music 

* Added audiomixer to Boss Room for separate audio channels and a master mixer. 

* Updated and added various hero abilities:

  * Added a cooldown to Archer's PowerShot. 
  * Used generic actions to implement Rogue's Dagger skill. 
  * Used generic actions to implement Rogue's Sneak skill using local stealth, applying a graphical effect to the stealthy character while still sending all network events normally. 
  * Used generic actions to implement Rogue's Dash skill. 
  * Properly display Heal abilities when targeting a fallen ally character. 
  * Character attack actions properly support Hold to charge options. 

> * To show how UI elements and game objects can be networked, added networked functionality using `INetworkSerializable` for UI elements on the Character Selection screen including networked mouse cursors. 

Author: Hmm gotcha, ticket provided that info, will revise. Thanks!

> * To show how UI elements and game objects can be networked, added networked functionality using `INetworkSerializable` for UI elements on the Character Selection screen including networked mouse cursors. 

Reviewer: Oh that's not great. I know it was mentioned, but was never cleaned up. Our fault.

> * Added a Photon filter for host room names not to allow profanity, racial slurs, and additional socially acceptable words. 

Reviewer: Just confirming, are names actually filtered for profanity? I know we had a task to introduce sanitization but that was mostly to filter out invisible characters introduced by Zoom when copy-pasting room names.

Author: Gotcha, the ticket was thin on details and didn't easily find into in PRs. Updating based on this. :)

Reviewer: Just checking. It seems this change remained.

* Boss Room now uses the [UnityToonShader](https://github.com/IronWarrior/UnityToonShader) for rendering the 3D surfaces to emulate 2D, flat surfaces.
* Added disconnection error message to load when a player or host disconnects due to limited or no network connectivity. Client logic was also updated to detect Host disconnection scenarios, such as losting connectivity.
* Balanced hero and enemy stats, spawners, trigger areas, and enemy detetction areas. 
* Added healthbars to display when damaged and properly track imp health locally and across clients. 

### Changes

* Updated the Photon Setup Guide, indicating you need only app ID when playing with friends. For users connecting across regions, you may need to hard code a region in your app settings by using the room code and region instead of just the room code sharing in game. 

> * Removed Singleton usage, allowing multiple instances of MLAPI networking stack to start up in the same process. 

Reviewer: Is this referring to MLAPI's `NetworkManager.Singleton`? We haven't yet updated to the newest version that strips the singleton, so we still have `NetworkManager.Singleton` references in BossBoom.

Author: Gotcha! thanks!

* On Wizards, fixed issue with imp spawners not respawning new imps after exploring the room.
* Fixed an issue where the door state does not reflect the existing state when players connect late to a game, for example if other players open the door and a player joins late the door displays as closed. 
* Removed a previous work-around for character selections when host replays a completed game. The issue was resolved, allowing players to see character selections during replay. 
* Fixed collision wall settings, fixing an issues where the boss knock-back ability sent players through walls.
* Resolved an issue where any players leaving the lobby sent all players to the lobby.
* Fixed animations for enemies including the smoke animation for destroyed imps and the boss helmet when crying.

### Known issues

* An MLAPI soft sync error on cleanup between scene transitions may break the game, for example imps do not spawn and pots are intangible.
* Sometimes after completing a match and the host starts a new match from the Victory or Loss screen, connected players may have no visible interactions to join or select characters. A work-around is implemented to not block entry into the game. 

> 

Reviewer: This was the other known issue in question: https://unity3d.atlassian.net/browse/GOMPS-534.

Author: Adding!

### New features

* Introduced static scene `NetworkObject`s to Boss Room including the following updates:

    * Added two separator `GameObject`s for scene readability: runtime `NetworkObject`s and `NetworkObject`s already placed in the scene.

>     * Added `GameEvent` (ScriptableObject) and `GameEventListener` (MonoBehaviour) to encapsulate events inside assets, located in the `ServerBossRoomState` prefab which now has a `GameEventListener` component. The event associated to this listener is `BossDefeated`, which the Boss raises when the `LifeState` is Dead in the `RaiseEventOnLifeChange` component. This feature can be expanded for other events, such as characters deaths.

Reviewer: I would just remove the last sentence here.
